### PR TITLE
Pushed filters to Parquet file on best effort basis in Vectorized Reader

### PR DESCRIPTION
**Pull request by mudit-97**
Hi all,

We are planning to onboard Iceberg as our Tableformat stack in our Data Platform. For that purpose, we were doing some benchmarking between Hudi and Iceberg vs Vanilla Parquet based Dataframe reads in spark and we found out that Iceberg is not pushing the filters to the parquet layer and it is relying on row group based filtering which is a customized pattern for Iceberg

However we saw that amount of data read will be higher in that case, we tried to push down filters till the parquet layer also and we saw benefits in some queries.

We wanted to know the comments on this PR whether this makes sense to the community or should this be avoided and rely on only row group filtering for Iceberg

### CODE REVIEW

**Comment by mudit-97** [comment]
@Fokko @nastra can you please check this once or tag relevant folks if possible

**Comment by gvramana** [comment]
@ajantha-bhat can you check and help on this

**Comment by ajantha-bhat** [comment]
Can we execute spotless `./gradlew spotlessApply` and also fix the test case failure? 

I haven't worked closely in this module of Iceberg. But I will take a look. I will also tag others for review.

**Comment by mudit-97** [comment]
@ajantha-bhat , ack, will fix UTs and comment here

**Comment by mudit-97** [comment]
@ajantha-bhat fixed the test cases and applied spotless

**Comment by ajantha-bhat** [comment]
I think still there are checkstyle failure, revAPI failure (since public method params changed) and a test case failure. 

Can also locally run `./gradlew clean build -x test -x integrationTest` to first have a clean build before pushing.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 1]
```diff
@@ -104,19 +104,21 @@
 import org.apache.parquet.column.ParquetProperties.WriterVersion;
 import org.apache.parquet.crypto.FileDecryptionProperties;
 import org.apache.parquet.crypto.FileEncryptionProperties;
-import org.apache.parquet.hadoop.ParquetFileReader;
-import org.apache.parquet.hadoop.ParquetFileWriter;
-import org.apache.parquet.hadoop.ParquetOutputFormat;
+import org.apache.parquet.hadoop.*;
```
Comment (ajantha-bhat) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:107]: if you check compile you will have checkstyle error for this. Iceberg doesn't use * import.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 2]
```diff
@@ -1160,8 +1162,37 @@ public <D> CloseableIterable<D> build() {
           optionsBuilder.withDecryption(fileDecryptionProperties);
         }
 
+        boolean pushedFilters = true;
         ParquetReadOptions options = optionsBuilder.build();
 
+        if (filter != null) {
```
Comment (ajantha-bhat) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:1168]: This is code duplication from below block. We can extract the common logic to form `ParquetFilters`

**Review by ajantha-bhat — COMMENTED** [review]
I spent little time and understood this PR. 

Basically you want to enable record level filtering (and you have observed the benefits with this POC PR) for vector reader instead of high level row-group filtering.

You want to know why it is not enabled for vector reader but enabled for plain ParquetReadBuilder without reader functions (readerFunc and batchedReaderFunc). 

I don't have answer for this history. I can check in iceberg slack or mailing list. 
We might have to run all benchmarks and understand whether any side effects of enabling record level filtering.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java (hunk 1 above)
Comment (mudit-97) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:107]: @ajantha-bhat yes I have fixed this in my local, fixing 1 Integration test case, then will push it again

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java (hunk 2 above)
Comment (mudit-97) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:1168]: ack

**Comment by ajantha-bhat** [comment]
@mudit-97: I did ask the question on iceberg slack (dev channel), feel free to join the channel. 
@amogh-jahagirdar pointed me to the historical PR for the same (https://github.com/apache/iceberg/pull/1566). 

I am yet to go through and summarise it.

**Comment by mudit-97** [comment]
@ajantha-bhat thanks for the help, I will also join the channel

**Pull request closed without merging**